Tutorial OIDC (#2339)
giantswarm/docs: src/content/tutorials/access-management/configure-dex-in-your-cluster/index.md

---
linkTitle: OIDC auth for workload clusters
title: Configure OIDC using Dex to access your clusters
description: A general explanation on how to install and configure Dex to work as an authenticator mechanism to provide OpenID tokens.
weight: 10
menu:
  principal:
    parent: tutorials
    identifier: tutorials-access-management
user_questions:
  - How can I configure OIDC in my cluster?
  - How can I add a new OIDC connector?
last_review_date: 2024-11-22
owner:
  - https://github.com/orgs/giantswarm/teams/team-bigmac
---

## Introduction

At Giant Swarm, we automatically configure Dex in management clusters to allow you to authenticate using your own identity providers, towards allowing you to manage your infrastructure using the management cluster's Kubernetes API.

> For workload clusters - where you run your applications - we do not enforce any specific OpenID Connect (OIDC) tool to enable single sign-on (SSO). Here, we're going to detail how to configure [Dex](https://dexidp.io/) in those clusters, to provide SSO using OIDC.

Author: ```suggestion
For workload clusters - where you run your applications - we don't enforce any specific OpenID Connect (OIDC) tool to enable single sign-on (SSO). Here, we're going to detail how to configure [Dex](https://dexidp.io/) in those clusters, to provide SSO using OIDC.
```

![Multi cluster Dex architecture](dex-architecture.png)
<!-- Source: https://drive.google.com/file/d/12Li9z2cqS8uWo1f9bGk6nwV6PLgty9g_ -->

## Why Dex

> There are other projects that help to configure OIDC to access Kubernetes clusters, but we consider [Dex](https://dexidp.io/) to be the most feature-rich. First of all, it is not tied to Kubernetes, so you can use Dex to handle authentication and authorization for your own apps as well. Secondly, Dex can act like an identity provider hub, where you can plug in different providers via different connectors, and choose between them when you want to log in.

Author: ```suggestion
There are other projects that help to configure OIDC to access Kubernetes clusters, but we consider [Dex](https://dexidp.io/) to be the most feature-rich. First of all, it's not tied to Kubernetes, so you can use Dex to handle authentication and authorization for your own apps as well. Secondly, Dex can act like an identity provider hub, where you can plug in different providers via different connectors, and choose between them when you want to log in.
```

> ## OIDC in Kubernetes

Author: ```suggestion
## OpenID Connect in Kubernetes
```

The Kubernetes API allows users to authenticate using the OIDC protocol, making it possible to enforce multi-factor authentication (MFA) or password policies by delegating to your identity provider. The API will use the field named `id_token` from the response as a bearer token to authenticate users.

> ## Configure the OIDC values on the cluster resource

Author: ```suggestion
## Configure the OpenID Connect values on the cluster resource
```

    spec:
      kubeadmConfigSpec:
        clusterConfiguration:
          apiServer:
            extraArgs:
              oidc-issuer-url: https://dex.CLUSTER_NAME.BASE_DOMAIN
              oidc-client-id: dex-k8s-authenticator
              oidc-username-claim: email
              oidc-groups-claim: groups
```

{{< /tab >}}
{{< tab title="Cluster API (any)">}}

```yaml
# Config map with values for the Workload Cluster app
apiVersion: v1
kind: ConfigMap
data:
  values: |
    ...
    global:
      controlPlane:
        oidc:
          issuerUrl: https://dex.CLUSTER_NAME.BASE_DOMAIN
          clientId: dex-k8s-authenticator
          usernameClaim: email
          groupsClaim: groups
```

{{< /tab >}}
{{< tab title="Cluster API (AWS EKS)">}}

```yaml
# Config map with values for the Workload Cluster app
apiVersion: v1
kind: ConfigMap
data:
  values: |
    ...
    global:
      controlPlane:
        oidcIdentityProviderConfig:
          issuerUrl: https://dex.CLUSTER_NAME.BASE_DOMAIN
          clientId: dex-k8s-authenticator
          usernameClaim: email
          groupsClaim: groups
          identityProviderConfigName: dex-k8s-authenticator
```

{{< /tab >}}
{{< /tabs >}}

> __Note__: In the above snippets you need to replace the `CLUSTER_NAME` and `BASE_DOMAIN` placeholder with the correct values, which is the name of the workload cluster you are configuring, and the base domain that you use for your installation. You can also derive them from the workload cluster's Kubernetes API endpoint, which has a URL in the format of `https://api.CLUSTER_NAME.BASE_DOMAIN`.

Author: ```suggestion
__Note__: In the above snippets you need to replace the `CLUSTER_NAME` and `BASE_DOMAIN` placeholder with the correct values, which is the name of the workload cluster you are configuring, and the base domain that you use for your installation. You can also derive them from the workload cluster's Kubernetes API endpoint, which has an address in the format of `https://api.CLUSTER_NAME.BASE_DOMAIN`.
```

## Deploy the app to your cluster

> In this guide, we will use a single app deployment for each cluster that you want to authenticate towards. There are different ways to set up how you authenticate towards your Kubernetes API with Dex, but in our opinion, using a single deployment per cluster is more resilient than having a common Dex deployment for all your workload clusters.

Author: ```suggestion
In this guide, you will use a single app deployment for each cluster that you want to authenticate towards. There are different ways to set up how you authenticate towards your Kubernetes API with Dex, but in our opinion, using a single deployment per cluster is more resilient than having a common Dex deployment for all your workload clusters.
```

> We'll use the [app platform]({{< relref "/vintage/getting-started/app-platform" >}}) to deploy the app, as it allows us to deploy apps across workload clusters using a single API endpoint. In this example, we create an `App` custom resource (CR) with the parameters to install our [`dex-app`](https://github.com/giantswarm/dex-app) in the desired cluster, and a `Secret` with the configuration values. The app platform also supports providing the configuration values in a `ConfigMap`, but in case of Dex some configuration values are credentials, therefore using a `Secret` is the preferred method. It is also possible to split the configuration values, provide the public part in a `ConfigMap` and the credentials in a `Secret`, but in this example we will provide the entire configuration in a `Secret`:

Author: ```suggestion
You will use the [app platform]({{< relref "/overview/app-platform" >}}) to deploy the app, as it allows you to deploy apps across workload clusters using a single API endpoint. In this example, you create an `App` custom resource (CR) with the parameters to install your [`dex-app`](https://github.com/giantswarm/dex-app) in the desired cluster, and a `Secret` with the configuration values. The app platform also supports providing the configuration values in a `ConfigMap`, but in the case of Dex, some configuration values are credentials; therefore, using a `Secret` is the preferred method. It's also possible to split the configuration values, provide the public part in a `ConfigMap` and the credentials in a `Secret`, but in this example, you will provide the entire configuration in a `Secret`:
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: dex-app-user-values
  namespace: CLUSTER_NAMESPACE_NAME
data:
  values: ... # base64 encoded connector config
```

> The format of the `connector config` content can look different depending on the oidc provider you want to use. Some examples can be found below.

Author: ```suggestion
The format of the `connector config` content can look different depending on the OpenID Connect provider you want to use. Some examples can be found below.
```

Details on all connectors and their respective configuration is available in the [Dex documentation](https://dexidp.io/docs/connectors/).

{{< tabs >}}
{{< tab title="Keycloak">}}

```yaml
isWorkloadCluster: true
oidc:
  expiry:
    signingKeys: 6h
    idTokens: 30m
  customer:
    enabled: true
    connectors:
    - id: customer
      connectorName: test
      connectorType: oidc
      connectorConfig: |
        clientID: CLIENT_ID
        clientSecret: CLIENT_SECRET
        insecureEnableGroups: true
        scopes:
        - email
        - groups
        - profile
        issuer: https://KEYCLOAK_HOST/realms/master
        redirectURI: https://dex.CLUSTER_NAME.BASE_DOMAIN/callback
```

> The values for `CLIENT_ID` and `CLIENT_SECRET` are created/defined in the Keycloak OIDC client configuration. `KEYCLOAK_HOST` is the fully qualified host name of your Keycloak instance.

Author: ```suggestion
The values for `CLIENT_ID` and `CLIENT_SECRET` are created/defined in the `Keycloak` OpenID Connect client configuration. `KEYCLOAK_HOST` is the fully qualified hostname of your `Keycloak` instance.
```

{{< /tab >}}
{{< tab title="GitHub">}}

```yaml
isWorkloadCluster: true
oidc:
  expiry:
    signingKeys: 6h
    idTokens: 30m
  customer:
    enabled: true
    connectors:
    - id: customer
      connectorName: test
      connectorType: github
      connectorConfig: |
        clientID: CLIENT_ID
        clientSecret: CLIENT_SECRET
        loadAllGroups: false
        teamNameField: slug
        redirectURI: https://dex.CLUSTER_NAME.BASE_DOMAIN/callback
        orgs:
        - name: GITHUB_ORGANIZATION
          teams:
          - GITHUB_TEAM_SLUG
```

> The value for `CLIENT_ID` is automatically generated by GitHub after creation of an OAuth app. Please also generate a client secret for that app and replace `CLIENT_SECRET` with the contents of that secret.

Author: ```suggestion
The value for `CLIENT_ID` is automatically generated by GitHub after the creation of an `OAuth` app. Please also generate a client secret for that app and replace `CLIENT_SECRET` with the contents of that secret.
```

> In the above example we show how to configure access for a certain GitHub team (named `GITHUB_TEAM_SLUG`, belonging to the `GITHUB_ORGANIZATION` organization). To restrict access in this way, make sure to list the team's _slug_ as it appears in the handle and team URL. For example, for a team named `Team Infra`, the handle my be `@my-organization/team-infra` and the slug to be used here would be `team-infra`.

Author: ```suggestion
In the above example you see how to configure access for a certain GitHub team (named `GITHUB_TEAM_SLUG`, belonging to the `GITHUB_ORGANIZATION` organization). To restrict access in this way, make sure to list the team's _slug_ as it appears in the handle and team URL. For example, for a team named `Team Infra`, the handle may be `@my-organization/team-infra`, and the slug used here would be `team-infra`.
```

        clientID: CLIENT_ID
        clientSecret: CLIENT_SECRET
        tenant: TENANT
        redirectURI: https://dex.CLUSTER_NAME.BASE_DOMAIN/callback
        useGroupsAsWhitelist: true
        groups:
          - AD_GROUP_NAME
```

The values to be filled in as `CLIENT_ID` and `CLIENT_SECRET` are obtained when creating a client application in Active Directory. Please replace `TENANT` with your Azure AD tenant ID.

Specifying `groups` and `useGroupsAsWhitelist` enables Dex's [group filtering]({{< relref "/vintage/advanced/access-management/configure-dex-in-your-cluster#dex-group-filtering" >}}).

{{< /tab >}}
{{< tab title="Okta">}}

```yaml
isWorkloadCluster: true
oidc:
  expiry:
    signingKeys: 6h
    idTokens: 30m
  customer:
    enabled: true
    connectors:
    - id: customer
      connectorName: test
      connectorType: oidc
      connectorConfig: |
        clientID: CLIENT_ID
        clientSecret: CLIENT_SECRET
        insecureEnableGroups: true
        getUserInfo: true
        scopes:
        - email
        - groups
        - profile
        issuer: https://OKTA_OIDC_ENDPOINT
        redirectURI: https://dex.CLUSTER_NAME.BASE_DOMAIN/callback
```

> The values for `CLIENT_ID` and `CLIENT_SECRET` have to be created in the Okta configuration. `OKTA_OIDC_ENDPOINT` must be replaced using the fully qualified host name of your Okta instance.

Author: ```suggestion
The values for `CLIENT_ID` and `CLIENT_SECRET` must be created in the `Okta` configuration. `OKTA_OIDC_ENDPOINT` must be replaced with the fully qualified host name of your `Okta` instance.
```

In case the cert-manager is configured to use a custom certification authority, which is not publicly trusted, Dex needs to be configured to trust it.
The custom certification authority needs to be stored in a `Secret` and referenced in the `trustedRootCA` section of the `Secret` or `ConfigMap` with configuration values for the Dex app.

```yaml
trustedRootCA:
  secretName: ... # Name of the secret, which contains the custom CA
  name: ... # Name of the property inside the secret
```

It is also possible to disable retrieving certificates from the cert-manager and supply custom certificates to Dex directly. They can be configured in the `ingress` section of the `Secret` or `ConfigMap` with the configuration values for the Dex app:

```yaml
ingress:
  tls:
    letsencrypt: false
    caPemB64: ...
    crtPemB64: ...
    keyPemB64: ...
```

### Running Dex in a private workload cluster

When deploying Dex to a private workload cluster, it is required to specify a proxy configuration in order to ensure that Dex has access to the outside network.
Proxy setup can be provided in the `cluster.proxy` section of the `Secret` or `ConfigMap` with configuration values for the Dex app.

> It is required to specify the URL of the HTTPS proxy in the `https` property and exclude Kubernetes API from running through the proxy by adding its IP address or network to the `noProxy` property.

Author: ```suggestion
It's required to specify the address of the `HTTPS` proxy in the `https` property and exclude Kubernetes API from running through the proxy by adding its IP address or network to the `noProxy` property.
```

```yaml
cluster:
  proxy:
    http: ... # HTTP proxy URL
    https: ... # HTTPS proxy URL
    noProxy: ... # Hostnames or networks/IP ranges excluded from going through the proxy

```

## Dex Observability

To get an overview on the authentication success and error rates of your Dex instances, we offer a Grafana dashboard named "Dex" as part of our [observability setup]({{< relref "/tutorials/observability/data-exploration/accessing-grafana" >}}).

## Logging into your workload cluster via `kubectl gs` with `Athena`

Once Dex is set up in your workload cluster, you can enable access via OIDC through our `kubectl` plugin [kubectl gs]({{< relref "/vintage/use-the-api/kubectl-gs" >}}).

> In order to communicate with the API, `kubectl gs` needs the clusters CA certificate as well as some cluster specific information, such as the management cluster name and the dex issuer URL.

Author: ```suggestion
In order to communicate with the API, `kubectl gs` needs the cluster's CA certificate as well as some cluster-specific information, such as the management cluster name and the dex issuer address.
```

```

__Note__: For workload cluster using [Cluster API `AWS`](https://github.com/giantswarm/cluster-aws) (CAPA) provider, use `443` instead of `6443` for the API address port.

__Warning__: For workload cluster using [Cluster API `EKS`](https://github.com/giantswarm/cluster-eks) provider, you'll need to configure Athena to use an AWS-managed EKS API server endpoint. This API server endpoint is uniquely allocated to your EKS cluster and can be easily accessed through the AWS EKS console by navigating to the `Overview` tab and under the `Details` section from the EKS cluster information page. For example:

```yaml
kubernetes:
  api:
    address: https://6EAE2F2E28XUD92EXZF54DFEF7C37081D.gr7.eu-central-1.eks.amazonaws.com
```

Access to Athena can be restricted to certain CIDRs.

```yaml
security:
  subnet:
    customer:
      public: x.x.x.x/x,x.x.x.x/x
      private: x.x.x.x/x
    restrictAccess:
      gsAPI: true
```

Athena also supports custom certificates, same as the Dex app.
The certificates need to be provided in the `ingress` section of the configuration:

```yaml
ingress:
  tls:
    letsencrypt: false
    caPemB64: ...
    crtPemB64: ...
    keyPemB64: ...
```

> If both `Dex` and `Athena` are configured correctly and you have installed `kubectl gs` on your machine, you should be able to create a kubectl context using the management API URL.

Author: ```suggestion
If both `Dex` and `Athena` are configured correctly and you have installed `kubectl gs` on your machine, you can create a kubectl context using the management API address.
```
